python_modules/dagster/dagster/_core/definitions/partition_mapping.py

```diff
@@ -116,3 +116,50 @@ def get_downstream_partitions_for_partition_range(
         upstream_partitions_def: PartitionsDefinition,
     ) -> PartitionKeyRange:
         raise NotImplementedError()
+
+
+@experimental
+class TimeIntervalPartitionMapping(PartitionMapping):
+    """
+    Allows mapping arbitrary time intervals.
+    """
+
+    def __init__(self, interval: timedelta, offset: Optional[timedelta] = None):
+        super().__init__()
+
+        offset = offset or timedelta()  # zero timedelta
+
+        self.interval = interval
+        self.offset = offset
+
+    def get_upstream_partitions_for_partition_range(
+        self,
+        downstream_partition_key_range: Optional[PartitionKeyRange],
+        downstream_partitions_def: Optional[PartitionsDefinition],
+        upstream_partitions_def: PartitionsDefinition,
+    ) -> PartitionKeyRange:
+        assert isinstance(downstream_partitions_def, TimeWindowPartitionsDefinition)
+        assert isinstance(upstream_partitions_def, TimeWindowPartitionsDefinition)
+
+        mapped_range = PartitionKeyRange(
```

Comment:
If I understand correctly, this could return partition_key values that aren't part of the upstream `PartitionsDefinition`.  A couple cases where this could happen:
- Both upstream and downstream are daily-partitioned, and the time lag is 2 hours.
- Both upstream and downstream are daily-partitioned, starting on "2022-06-06", and we're looking for the upstream partitions for downstream partition "2022-06-06".